File: Hard/980/980.js
// 980. Unique Paths III
// You are given an m x n integer array grid where grid[i][j] could be:
// 1 representing the starting square. There is exactly one starting square.
// 2 representing the ending square. There is exactly one ending square.
// 0 representing empty squares we can walk over.
// -1 representing obstacles that we cannot walk over.
// Return the number of 4-directional walks from the starting square to the
// ending square, that walk over every non-obstacle square exactly once.

const DIRECTIONS = [
  [-1, 0],
  [0, 1],
  [1, 0],
  [0, -1],
];

const EMPTY = 0;
const START = 1;
const END = 2;
const OBSTACLE = -1;

/**
 * @param {number[][]} grid
 * @param {number} m
 * @param {number} n
 * @return {number[]}
 */
function findStart(grid, m, n) {
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] == START) return [i, j];
    }
  }
}

function countEmptySpaces(grid, m, n) {
  let count = 0;
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] == EMPTY) count++;
    }
  }
  return count;
}

/**
 * @param {number[][]} grid
 * @param {number} m
 * @param {number} n
 * @param {number} i
 * @param {number} j
 * @param {number} r
 * @param {number[]} total
 * @return
 */
function evaluateFourDirections(grid, m, n, r, c, e, total) {
  for (let k = 0; k < 4; k++) {
    let i = r + DIRECTIONS[k][0];
    let j = c + DIRECTIONS[k][1];
    if (i >= 0 && i < m && j >= 0 && j < n) {
      if (grid[i][j] != START && grid[i][j] != OBSTACLE) {
        if (grid[i][j] == END && e == 0) {
          total[0]++;
        }
        if (grid[i][j] == 0) {
          let temp = grid[i][j];
          grid[i][j] = OBSTACLE;
          evaluateFourDirections(grid, m, n, i, j, e - 1, total);
          grid[i][j] = temp;
        }
      }
    }
  }
}

/**
 * @param {number[][]} grid
 * @return {number}
 */
var uniquePathsIII = function (grid) {
  let m = grid.length;
  let n = grid[0].length;
  let [r, c] = findStart(grid, m, n);
  let e = countEmptySpaces(grid, m, n);
  let total = [0];
  evaluateFourDirections(grid, m, n, r, c, e, total);
  return total[0];
};

console.log("It's Showtime Folks!");

let grid = [
  [1, 0, 0, 0],
  [0, 0, 0, 0],
  [0, 0, 2, -1],
];
let result = uniquePathsIII(grid);
console.log(`Result: ${result}`);

File: hard/980/980.cpp
// 980. Unique Paths III
// You are given an m x n integer array grid where grid[i][j] could be:
// 1 representing the starting square. There is exactly one starting square.
// 2 representing the ending square. There is exactly one ending square.
// 0 representing empty squares we can walk over.
// -1 representing obstacles that we cannot walk over.
// Return the number of 4-directional walks from the starting square to the
// ending square, that walk over every non-obstacle square exactly once.

#include <iostream>
#include <utility>
#include <vector>

using namespace std;

const int DIRECTIONS[4][2] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
const int EMPTY = 0;
const int START = 1;
const int END = 2;
const int OBSTACLE = -1;

pair<int, int> findStart(vector<vector<int>> &grid, pair<int, int> &size) {
  for (int i = 0; i < size.first; i++)
    for (int j = 0; j < size.second; j++)
      if (grid[i][j] == START)
        return {i, j};
  return {-1, -1};
}

int countEmptyCells(vector<vector<int>> &grid, pair<int, int> &size) {
  int count = 0;
  for (int i = 0; i < size.first; i++)
    for (int j = 0; j < size.second; j++)
      if (grid[i][j] == EMPTY)
        count++;
  return count;
}

// DFS (Depth-First Search) vs BFS (Breath-First Search)
void dfs(vector<vector<int>> &grid, pair<int, int> &size, pair<int, int> coords,
         int empty, int &total) {
  int i, j;
  for (int k = 0; k < 4; k++) {
    i = coords.first + DIRECTIONS[k][0];
    j = coords.second + DIRECTIONS[k][1];
    if (i >= 0 && i < size.first && j >= 0 && j < size.second)
      if (grid[i][j] != START && grid[i][j] != OBSTACLE) {
        if (grid[i][j] == 2 && empty == 0)
          total++;
        if (grid[i][j] == 0) {
          // cout << "(" << i << "," << j << ")" << endl;
          int temp = grid[i][j];
          grid[i][j] = OBSTACLE;
          dfs(grid, size, {i, j}, empty - 1, total);
          grid[i][j] = temp;
        }
      }
  }
}

class Solution {
public:
  int uniquePathsIII(vector<vector<int>> &grid) {
    pair<int, int> size = {grid.size(), grid[0].size()};
    pair<int, int> coords = findStart(grid, size);
    int empty = countEmptyCells(grid, size);
    int total = 0;
    dfs(grid, size, coords, empty, total);
    return total;
  }
};

int main() {
  cout << "It's Showtime Folks" << endl;
  vector<vector<int>> grid = {{1, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 2, -1}};
  cout << "Solution: " << Solution().uniquePathsIII(grid) << endl;
  return 0;
}
